AstSemantics.md

```diff
@@ -32,35 +32,62 @@ Callstack space is limited by unspecified and dynamically varying constraints.
 a trap occurs.
 
-## Local and Memory types
-
-Individual storage locations in WebAssembly are typed, including global
-variables, local variables, and parameters. The main storage of a wasm module, 
-called the *Linear memory*, is a contiguous, byte-addressable range of memory
-spanning from offset `0` to `memory_size`. The linear memory can be considered 
-to be a untyped array of bytes which is stored separately from the global variables.
-The linear memory is sandboxed; it does not alias the execution engine's internal
-data structures, the execution stack, or other process memory.
-All accesses to memory are annotated with a type. The legal types for
-global variables and memory accesses are called *Memory types*.
+## Types
 
-  * `int8`: 8-bit integer
-  * `int16`: 16-bit integer
-  * `int32`: 32-bit integer
-  * `int64`: 64-bit integer
-  * `float32`: 32-bit floating point
-  * `float64`: 64-bit floating point
+### Basic Types
 
-The legal types for parameters and local variables, called *Local types*
-are a subset of the Memory types:
+The following types are called the *basic types*:
 
   * `int32`: 32-bit integer
   * `int64`: 64-bit integer
   * `float32`: 32-bit floating point
   * `float64`: 64-bit floating point
 
-All operations except loads and stores deal with local types. Loads convert
-Memory types to Local types according to the following rules:
+Note that the basic types `int32` and `int64` are not inherently signed or
+unsigned. The interpretation of these types is determined by individual
+operations.
+
+Parameters and local variables use basic types.
+
+### Expression Types
+
+*Expression types* include all the basic types, and also:
+
+  * `void`: no value
+
+AST expression nodes use expression types.
+
+### Memory Types
+
+*Memory types* are a different superset of the basic types, adding the
+following:
+
+  * `int8`: 8-bit integer
+  * `int16`: 16-bit integer
+
+Global variables and linear memory accesses use memory types.
+
+## Linear Memory
+
+The main storage of a wasm module, called the *linear memory*, is a contiguous,
+byte-addressable range of memory spanning from offset `0` and extending for
+`memory_size` bytes. The linear memory can be considered to be a untyped array
+of bytes. The linear memory is sandboxed; it does not alias the execution
+engine's internal data structures, the execution stack, local variables, global
+variables, or other process memory.
+
+In the MVP, linear memory is not shared between threads. When
+[threads](PostMVP.md#threads) are added as a feature, regular load and store
+nodes will have the most relaxed semantics specified in the memory model and new
+memory-access nodes will be added with atomic and ordering guarantees.
+
+### Linear Memory Operations
+
+Linear memory operations are annotated with a memory type and perform a
+conversion between that memory type and a basic type.
+
+Loads read data from linear memory, convert from their memory type to a basic
```

Comment:
I recall using "local type" in AstSemantics. Maybe those references have
been refactored away. "Local" is nice because it relates to local variables.

On Tue, Jul 7, 2015 at 2:52 PM, Dan Gohman <notifications@github.com> wrote:

> In AstSemantics.md
> <https://github.com/WebAssembly/design/pull/245#discussion_r34035409>:
>
> > +`memory_size` bytes. The linear memory can be considered to be a untyped array
> > +of bytes. The linear memory is sandboxed; it does not alias the execution
> > +engine's internal data structures, the execution stack, local variables, global
> > +variables, or other process memory.
> > +
> > +In the MVP, linear memory is not shared between threads. When
> > +[threads](PostMVP.md#threads) are added as a feature, regular load and store
> > +nodes will have the most relaxed semantics specified in the memory model and new
> > +memory-access nodes will be added with atomic and ordering guarantees.
> > +
> > +### Linear Memory Operations
> > +
> > +Linear memory operations are annotated with a memory type and perform a
> > +conversion between that memory type and a basic type.
> > +
> > +Loads read data from linear memory, convert from their memory type to a basic
>
> I grepped the design repo and didn't see any other references. Do you
> prefer "local" over "basic" here? These types do get used outside of
> "local" contexts.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/WebAssembly/design/pull/245/files#r34035409>.
>